### DISCUSSION TITLE
How to make kwok controller handle pod faster

### DISCUSSION THREAD

**Discussion by jokerwyt** [kubernetes-sigs/kwok#1273] [Q&A] [answered]
I need to create 2000 pods per second for benchmarking and I found that kwok controller cannot assign pod ready status so fast. What can I do to make it catch up the speed I create pods?

**Comment by wzshiming** [comment]
I suspect your bottleneck is that the kube-scheduler isn't fast enough

You can try the following steps to speed things up
- Fill the nodeName of pod to bypass kube-scheduler
- Twisted kube-apiserver qps limits on client's connections

**Reply by jokerwyt** [reply]
Thanks! But I am a relatively experienced user of K8s, and I have already done what you said before I came up with that question. All components' QPS are twisted to 100000, I think it's enough for my use case. Kube-scheduler is also bypassed. I also tried to change `pod-play-stage-parallelism` from 4 to 16. It mitigates the problem, but not fully fixed.

I wonder a method to determine whether KWOK is fast enough to watch pods and put Ready Condition on it, like some prometheus workqueue metrics?

**Reply by wzshiming** [reply]
Oh, you are the second user with this requirement

The result of the discussion at that time was [a kwok controller managing only one node](https://github.com/Azure/kperf/blob/6b4fce21d64e0d49e233c280aff51b24fd75fd15/manifests/virtualcluster/nodecontrollers/templates/statefulsets.tpl#L37) as a workaround.

because, [kube-apiserver limits the client's qps](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/#queuing), and twisting it up doesn't help too much.

I recently had a new idea of this, [creating multiple connections directly in a single controller](https://github.com/wzshiming/kwok/commit/786d7a6bce81533f1f0e30028436221a54b1a8fe), but I haven't actually tested its performance yet.

**Reply by jokerwyt** [reply]
Em, I don't see the difference between managing 1 kwok node and 100 kwok nodes (my case). But I think multiple connections can be better than a single one. 

I have just confirmed that KWOK can handle at least 1000 pod stage changes per second, by using another controller to watch pods' status change directly (with `pod-play-stage-parallelism`=64). It shows that the pods ready number can catch up with the pod creation event.  When I tried to scale up, the pod creation became the bottleneck.

**Reply by wzshiming** [reply]
> I have just confirmed that KWOK can handle at least 1000 pod stage changes per second, by using another controller to watch pods' status change directly (with pod-play-stage-parallelism=64). It shows that the pods ready number can catch up with the pod creation event. When I tried to scale up, the pod creation became the bottleneck.

The current default on my machine and CI, it's processing speed is comparable to pod creation. What version and environment do you use?

**Reply by jokerwyt** [reply]
k8s: v1.29.1
kwok controller image: v0.6.0-126-gf972843
Ubuntu 22.04.2

**Reply by jokerwyt** [reply]
I think a larger `pod-play-stage-parallelism` default value can fix this. I expose it as a command line arg of the kwok controller. If needed I can contribute [my changes](https://github.com/jokerwyt/kwok) to this repo.

**Reply by wzshiming** [reply]
Your contribution is welcome. 

I would expect to see significant improvements in CI for performance optimizations.

This is the current benchmark.

#1276
https://github.com/kubernetes-sigs/kwok/actions/runs/11967429145/job/33364640178?pr=1276

**Reply by jokerwyt** [reply]
For your reference, I did a simple benchmark (create pods with a speed of 1400 pods per second) and checked some logs of the replica set controller. 
It observed pod creation (green line) with a higher speed than pod-ready condition (orange line, with `pod-play-stage-parallelism=64`). I think it implies that kwok's stage playing is becoming a bottleneck (under such a not-realistic case though)

![image](https://github.com/user-attachments/assets/f70e61c5-f6f8-4aa9-953a-d44d2bcddd84)

**Reply by wzshiming** [reply]
I wonder what this image was made of. Is it possible to compare different parallel numbers?

**Reply by jokerwyt** [reply]
It was made by the log of kube-controller-manager (i.e. the replicaset controller). That controller watches pod creation and pod ready status to calculate replicaset replicas number and ready replicas number. 

Is it possible to compare different parallel numbers? Yeah I will do that comparison if I needed, and post the result here.

**Reply by wzshiming** [reply]
So that means kube-controller-manager creates pods faster than kwok updates them...

**Reply by wzshiming** [reply]
Is the kwok-controller on the same node as the kube-apiserver?

**Reply by jokerwyt** [reply]
Yes. I suspect that ETCD is the bottleneck. It cannot handle such high rate of POST and UPDATE mix workload

**Reply by jokerwyt** [reply]
I introduced two Prometheus metrics to monitor how fast kwok-controller preprocesses and playstage. Here is the result when creating 1k pods per second for 1 minute (4:38:45~4:39:45). I don't know why the kwok controller playstage speed rises after 4:39:45. It does not look like cpu pressure because overall machine CPU usage is about 60%. 

One possible explanation is ETCD performs better after 4:39:45 because there is no pod creation anymore. 

![image](https://github.com/user-attachments/assets/194ba941-b765-440c-8db1-bd367bf25a85)
![image](https://github.com/user-attachments/assets/3455afee-f276-457d-aaee-59972e70f1dc)

**Comment by wzshiming** [comment]
I plan to find the bottleneck and fix it. https://github.com/kubernetes-sigs/kwok/issues/1283

**Comment by ali-a-a** [comment]
@jokerwyt

I had the same issue with KWOK as I wanted to load my K8s cluster using [CL2](https://github.com/kubernetes/perf-tests/tree/master/clusterloader2). My test was creating 2000 deployments with 30 replicas at a rate of 50 deployments per second. I faced many `PendingScheduled` pods waiting for KWOK to change their status to `Running`. I worked around this by running multiple KWOK processes, each managing some worker nodes using annotation selectors. In my case, having 4 KWOK processes could catch up with K8s components.

**Reply by jokerwyt** [reply]
Yeah, I think this proves that Kwok processes is a very possible bottleneck.

**Comment by wzshiming** [comment]
I've been busy lately, and I'll focus on improving performance in the next release.

**Comment by wzshiming** [comment, marked as answer]
Sorry for taking so long to reply.

Lately I have had some time to devote to this project, I have been debugging the root cause which is QPS limiting apiserver to kwok-controller.

I've tested that increasing concurrency doesn't improve much, adjust APF configuration for maximum improvement #1341
